### PR TITLE
[INLONG-3387][SDK] Strengthen the strategy of heartbeat and channel update for dataproxy

### LINKED ISSUE #3387: [Improve][SDK] Strengthen the strategy of heartbeat and channel update for dataproxy
### Description

_No response_

### InLong Component

InLong SDK

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by pocozh**
### Title Name: [INLONG-3387][dataproxy-sdk] Strengthen the strategy of heartbeat and channel update

where *XYZ* should be replaced by the actual issue number.

Fixes #3387 

### Modifications

- send heartbeat for idle dataChannel to keep alive
- reuse bad host while dataproxy hosts are rare, avoid there is no dataChannel
- improve code style

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Review by baomingyu — APPROVED** [review]
+1

File: inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java [hunk 1]
```diff
@@ -572,29 +577,40 @@ public int compare(Map.Entry<HostInfo, Integer> o1, Map.Entry<HostInfo, Integer>
...
+            if (sender.isIdleClient(clientEntry.getValue())) {
+                sendHeartBeat(clientEntry.getKey(), clientEntry.getValue());
+            }
+        }
+
+    }
+
+    private void sendHeartBeat(HostInfo hostInfo, NettyClient client) {
+        try {
+            if (client.isActive()) {
```
Comment (luchunliang) [inline inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java:601]: It is better to add a log when client is not active.

File: inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java [hunk 2]
```diff
@@ -572,29 +577,40 @@ public int compare(Map.Entry<HostInfo, Integer> o1, Map.Entry<HostInfo, Integer>
...
+                logger.debug("active host to send heartbeat! {}", hostInfo.getReferenceName());
+                EncodeObject encodeObject = new EncodeObject("heartbeat".getBytes(StandardCharsets.UTF_8),
+                        8, false, false, false, System.currentTimeMillis() / 1000, 1, "", "", "");
+                if (configure.isNeedAuthentication()) {
```
Comment (luchunliang) [inline inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java:605]: Heartbeat data miss HostInfo data.

File: inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java [hunk 3]
```diff
@@ -650,6 +666,25 @@ private void fillUpWorkClientWithLastBadClient() {
         }
     }
 
+    private void fillUpWorkClientWithBadClient(List<HostInfo> badHostLists) {
+        if (badHostLists.isEmpty()) {
```
Comment (luchunliang) [inline inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java:670]: It is good thing that badHostLists is empty, why use warn level log?

**Review by luchunliang — APPROVED** [review]

File: inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/config/ProxyConfigManager.java [hunk 4]
```diff
@@ -598,21 +606,21 @@ private ProxyConfigEntry getLocalProxyListFromFile(String filePath) throws Excep
             if (jsonItem != null) {
                 if (!jsonItem.has("port")) {
                     throw new Exception("Parse local proxyList failure: "
-                        + "port field is not exist in address(" + i + ")!");
+                            + "port field is not exist in address(" + i + ")!");
```
[the remark this answers is not in the document]
Comment (pocozh) [reply inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/config/ProxyConfigManager.java:609]: Using tencent-code-style template, long line break is 8 space

[the remark this answers is not in the document]
Comment (pocozh) [reply inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/config/ProxyConfigManager.java:609]: > Why is indentation 8 space?

Using tencent-code-style template, long line break is 8 space

File: inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java (hunk 3 above)
Comment (pocozh) [reply inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java:670]: > It is good thing that badHostLists is empty, why use warn level log?

Invoking this method means that all hosts are bad, so badHostLists should not be empty

File: inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java (hunk 1 above)
Comment (pocozh) [reply inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java:601]: done

Comment (pocozh) [reply inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java:601]: > It is better to add a log when client is not active.

done

File: inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java (hunk 2 above)
Comment (pocozh) [reply inlong-sdk/dataproxy-sdk/src/main/java/org/apache/inlong/sdk/dataproxy/network/ClientMgr.java:605]: > Heartbeat data miss HostInfo data.

done

**Review by dockerzhang — APPROVED** [review]

**Pull request merged**